keps/sig-storage/20200116-node-failover.md

```diff
@@ -0,0 +1,230 @@
+
+This behavior was very racy, as while the node was being deleted, it could come back and register itself, list its pods and start the containers while volumes are being detached, leading to data corruption. To mitigate this, we explicitly set requirements on handling the node objects for cloud providers[3], i.e. not deleting node object unless the instance do not exist anymore.
+
+This made it noticeable to users, several were complaining about it, as a temporary mitigation we introduced the node-shutdown taint, enabling cluster administrators to write automation against it, although most of the solutions at the time were racy.
+
+Solving this issue requires also taking into account the existing feature set, the most notable one is the alpha bootstrap checkpointing[4], but since it’s alpha it is not enabled by default which makes the kubelet use usual flow (i.e check in to get the latest state from the apiserver). Also, sig-cluster-lifecycle (which is its main consumer) decided to not pursue this path and start its deprecation[5]
+
+[1] [Pod safety design](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/pod-safety.md)
+[2] [Gc orphaned pods](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/gc_controller.go#L97)
+[3] [requirements on node objects](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L167)
+[8] [bootstrap checkpointing](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/0004-bootstrap-checkpointing.md)
+[9] [bootstrap checkpointing deprecation intent](https://github.com/kubernetes/enhancements/issues/1103)
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Introduce a new API to report node state
+* Handle cases where the Kubelet itself is unresponsive
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is shutdown without any intervention
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP plans to introduce the following capability to the CSI capability to the spec:
+
+```
+type PluginCapability_Service_Type int32
+
+const (
+	// SAFE_DETACH indicates that the volumes targeting
+	// this plugin MAY be safely detached from the node.
+	// If this capability is advertised The CO can assume
+	// that the volume won't be detached if it is still
+	// mounted and accessible from the container
+	PluginCapability_Service_SAFE_DETACH PluginCapability_Service_Type = 3
```

Comment:
Since PluginCapability_Service_Type is a generated type, "message PluginCapability" should be mentioned here:

```
// Specifies a capability of the plugin.
message PluginCapability {
  message Service {
    enum Type {
      UNKNOWN = 0;
      CONTROLLER_SERVICE = 1;
      VOLUME_ACCESSIBILITY_CONSTRAINTS = 2;
	// SAFE_DETACH indicates that the volumes targeting
	// this plugin MAY be safely detached from the node.
	// If this capability is advertised The CO can assume
	// that the volume won't be detached if it is still
	// mounted and accessible from the container
      SAFE_DETACH = 3;
    }
    Type type = 1;
  }
```